### PR TITLE
`Development`: Utilize server endpoint for getting longest working time in exam assessment

### LINKED ISSUE #8673: Change longest working time method with server endpoint
### Is your feature request related to a problem?

In exam-assessment-buttons.component.ts Line 69, it is stated that “The longest working time should be directly loaded from the server and loading of all student exams should be avoided.” An endpoint is implemented for this in PR #8120, named as findLongestWorkingTimeForExam in file [StudentExamRepository.java](https://github.com/ls1intum/Artemis/pull/8120/files#diff-267803625aa154b703fc87901fb4fc52289d0bb985e33432792ebf890dbe1a3c).  The mapping is done in getLongestWorkingTimeForExam in [StudentExamResource.java](https://github.com/ls1intum/Artemis/pull/8120/files#diff-8098ab9f22aba5ddb5a01571b4841f329ac30ab0ff3bce8d32223f8b1c991b6f) classes.

### Describe the solution you'd like

The classes should be find that is using this method and the loading of exams should be changed with the endpoint. After every change, the changes should be tested manually and automatically.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by iyannsch**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [ ] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [ ] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.

### Motivation and Context

Fixes #8673 

The server offers a new endpoint that allows to reduce the computational load on the client. This PR implements the changes in the client side.

### Description

I removed getting all exams and calculating the longest working time iteratively by a simple call to the new endpoint.

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Switch to a instructor account during the working time and open the `Assessment` tab of the exam.
4. Verify that the following buttons are visible 
image
5. Verify that only the `Grading Key` is accessible during the exam time.
6. Verify that all buttons are accessible after the exam.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

**Review by Strohgelaender — APPROVED** [review]

File: src/test/javascript/spec/component/assessment-dashboard/exam-assessment-buttons.component.spec.ts [hunk 1]
```diff
@@ -59,12 +58,9 @@
             },
         }),
         MockProvider(StudentExamService, {
-            findAllForExam: () => {
+            getLongestWorkingTimeForExam: () => {
                 return of(
```
[the remark this answers is not in the document]
Comment (iyannsch) [reply src/test/javascript/spec/component/assessment-dashboard/exam-assessment-buttons.component.spec.ts:62]: More consistent to rest of code if it is like this

[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/test/javascript/spec/component/assessment-dashboard/exam-assessment-buttons.component.spec.ts:62]: Never seen such a aggressive tone from coderabbit, wonder how that happened.

**Review by Strohgelaender — APPROVED** [review]

**Comment by harjotgill** [comment]
@coderabbitai configuration

**Review by marlon-luca-bu — APPROVED** [review]
Code 👍

File: src/main/webapp/app/course/dashboards/assessment-dashboard/exam-assessment-buttons/exam-assessment-buttons.component.ts [hunk 2]
```diff
@@ -64,18 +64,24 @@ export class ExamAssessmentButtonsComponent implements OnInit {
...
-                    this.longestWorkingTime = Math.max.apply(
-                        null,
-                        this.studentExams.map((studentExam) => studentExam.workingTime),
-                    );
+
+            /*
+            Prepare workingTimeObservable to perform the following on subscribe:
+            - set the longestWorkingTime
+            - trigger (re)calculation of whether the exam is over
+             */
```
Comment (florian-glombik) [inline src/main/webapp/app/course/dashboards/assessment-dashboard/exam-assessment-buttons/exam-assessment-buttons.component.ts:72]: This is a personal opinion, but I do think that the comments are not needed - for me it seems like a reformulation of the code that follows, which to me, seems to be easy to understand and does not need further explanation.

For me, the comment doesn't add information 😅

File: src/main/webapp/app/course/dashboards/assessment-dashboard/exam-assessment-buttons/exam-assessment-buttons.component.ts [hunk 3]
```diff
@@ -64,18 +64,24 @@ export class ExamAssessmentButtonsComponent implements OnInit {
...
+                    this.longestWorkingTime = value;
                     this.calculateIsExamOver();
                 }),
             );
 
+            /*
+            Prepare workingTimeObservable to perform the following on subscribe:
+            - set the exam
+            - trigger (re)calculation of whether the exam is over
+             */
```
Comment (florian-glombik) [inline src/main/webapp/app/course/dashboards/assessment-dashboard/exam-assessment-buttons/exam-assessment-buttons.component.ts:84]: I think there is a mistake in the comment, I would be in favor of removing the comment

```suggestion
            /*
            Prepare examObservable to perform the following on subscribe:
            - set the exam
            - trigger (re)calculation of whether the exam is over
             */
```

**Review by florian-glombik — CHANGES_REQUESTED** [review]
I think it would be better to remove the javadoc comments. I do not think that they add value here and are most likely to get outdated with the actual code, which might lead to misleading assumptions.

File: src/main/webapp/app/course/dashboards/assessment-dashboard/exam-assessment-buttons/exam-assessment-buttons.component.ts (hunk 2 above)
Comment (iyannsch) [reply src/main/webapp/app/course/dashboards/assessment-dashboard/exam-assessment-buttons/exam-assessment-buttons.component.ts:72]: You got a fair point with the last two lines of comment, but personally I think that the first line provides value. It describes the fact that the following code does not actually retrieve any code but rather only modifies the "query" for later evaluation. Taking that into consideration, I´d argue that the last two lines are required to distinguish the two query-preparation statements. Dont you think so?

**Review by marlon-luca-bu — APPROVED** [review]
Reapprove

**Review by sarpsahinalp — APPROVED** [review]
code lgtm

**Review by krusche — APPROVED** [review]

**Pull request merged**